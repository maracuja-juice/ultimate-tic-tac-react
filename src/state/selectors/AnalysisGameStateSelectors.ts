import { AnalysisGame, AppState, Player } from '../AppState';

export const getAnalysisGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      id: finishedGame.id!,
      board: finishedGame.gameState,
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
    };
  }

  return undefined;
};
