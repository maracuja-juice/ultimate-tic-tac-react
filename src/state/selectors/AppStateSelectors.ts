import { AnalysisGame, AppState, Player } from '../AppState';

export const getBoards = (state: AppState) => state.currentGame.board;

export const getCurrentPlayer = (state: AppState) =>
  state.currentGame.game.currentPlayer;

export const getWinningPlayer = (state: AppState) =>
  state.currentGame.game.winningPlayer;

export const getMoves = (state: AppState) => state.currentGame.moves;

export const getFinishedGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
      board: finishedGame.gameState,
      id: finishedGame.id!,
    };
  }

  return undefined;
};
