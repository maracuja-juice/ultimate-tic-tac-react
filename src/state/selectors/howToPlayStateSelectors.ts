import { createSelector } from 'reselect';
import { getHowToPlay } from './appStateSelectors';
import { HowToPlay } from '../AppState';
import { howToPlaySteps } from '../howToPlay/howToPlaySteps';

export const getCurrentHowToPlayText = createSelector(
  [getHowToPlay],
  (howToPlay: HowToPlay) => {
    return howToPlaySteps[howToPlay.stepNumber].text;
  },
);
