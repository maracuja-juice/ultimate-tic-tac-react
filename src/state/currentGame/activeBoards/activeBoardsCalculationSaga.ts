import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction } from '../../AppState';
import { CALCULATE_ALLOWED_BOARDS, setAllowedBoards } from './activeBoardsActions';
import { getBoards } from '../../selectors/AppStateSelectors';
import { getNewActiveBoards } from '../../../util/ActiveBoards';

function* calculateActiveBoards( action: GenericAction ) {
    const boards = yield select( getBoards );
    const lastMove = action.payload;

    const activeBoards = getNewActiveBoards(lastMove, boards);

    yield put( setAllowedBoards( activeBoards ) );
}

function* activeBoardsCalculationSaga() {
    yield takeEvery( CALCULATE_ALLOWED_BOARDS, calculateActiveBoards );
}

export default activeBoardsCalculationSaga;