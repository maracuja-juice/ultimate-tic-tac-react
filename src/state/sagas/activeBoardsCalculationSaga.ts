import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction, SmallBoardInformation, TileValue } from '../AppState';
import { arePointsEqual } from '../../util';
import { CALCULATE_ALLOWED_BOARDS, setAllowedBoards } from '../activeBoards/activeBoardActions';
import { getBoards } from '../selectors/AppStateSelectors';

function* calculateActiveBoards( action: GenericAction ) {
    const boards = yield select( getBoards );
    const lastMove = action.payload;

    let activeBoards = [lastMove];
    const boardLastMovePointsTo = boards.find( ( board: SmallBoardInformation ) =>
                                                   arePointsEqual( board.position, lastMove ) );
    const boardIsFinished = boardLastMovePointsTo.value !== TileValue.Empty;
    if (boardIsFinished) {
        const allUnfinishedBoards = boards.filter(
            ( board: SmallBoardInformation ) =>
                board.value === TileValue.Empty
        );
        activeBoards = allUnfinishedBoards.map( ( board: SmallBoardInformation ) => board.position );
    }

    yield put( setAllowedBoards( activeBoards ) );
}

function* activeBoardsCalculationSaga() {
    yield takeEvery( CALCULATE_ALLOWED_BOARDS, calculateActiveBoards );
}

export default activeBoardsCalculationSaga;