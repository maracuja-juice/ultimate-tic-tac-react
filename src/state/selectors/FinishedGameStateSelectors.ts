import { createSelector } from 'reselect';
import { getBoards, getIsReplay, getMoves, getWinningPlayer } from './AppStateSelectors';

export const getWinningPlayerAsString = createSelector( [getWinningPlayer], (winningPlayer => {
    if (winningPlayer === 0) {
        return 'X';
    } else if (winningPlayer === 1) {
        return 'Y';
    } else if (winningPlayer === null) {
        return null;
    } else {
        return undefined;
    }
}) );

export const getFinishedGameData = createSelector(
    [getWinningPlayerAsString, getIsReplay, getBoards, getMoves],
    ( winningPlayer, isReplay, boards, moves ) => {
        return {
            winner: winningPlayer,
            gameState: boards,
            isReplay,
            moves
        };
    } );