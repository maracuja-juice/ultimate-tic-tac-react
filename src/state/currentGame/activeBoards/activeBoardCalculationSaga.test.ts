import { expectSaga } from 'redux-saga-test-plan';
import { select } from 'redux-saga/effects';
import { getMoves } from '../../selectors/AppStateSelectors';
import {
  CALCULATE_ACTIVE_BOARDS,
  SET_ACTIVE_BOARDS,
} from './activeBoardsActions';
import activeBoardsCalculationSaga from './activeBoardsCalculationSaga';
import {
  activeBoardsForBoardWithThreeMovesMock,
  movesForBoardWithThreeMovesMock,
  movesForCircleFinishedBoardMock,
  movesForUnfinishedBoardMock,
} from '../../../__mocks__';

describe('activeBoardCalculationSaga', () => {
  it('should dispatch setActiveBoards Action, normal case', () => {
    return expectSaga(activeBoardsCalculationSaga)
      .provide([[select(getMoves), movesForBoardWithThreeMovesMock]])
      .put({
        type: SET_ACTIVE_BOARDS,
        payload: activeBoardsForBoardWithThreeMovesMock,
      })
      .dispatch({ type: CALCULATE_ACTIVE_BOARDS })
      .silentRun();
  });

  it(
    'should set activeBoards to all unfinished small boards ' +
      'when small board the last move points to is finished',
    () => {
      // these values depend on the unfinished board mock.
      return expectSaga(activeBoardsCalculationSaga)
        .provide([[select(getMoves), movesForUnfinishedBoardMock]])
        .put({
          type: SET_ACTIVE_BOARDS,
          payload: [
            { x: 0, y: 2 },
            { x: 1, y: 0 },
            { x: 1, y: 2 },
            { x: 2, y: 0 },
            { x: 2, y: 2 },
          ],
        })
        .dispatch({ type: CALCULATE_ACTIVE_BOARDS })
        .silentRun();
    },
  );

  it('should set activeBoards to nothing when whole board is finished', () => {
    return expectSaga(activeBoardsCalculationSaga)
      .provide([[select(getMoves), movesForCircleFinishedBoardMock]])
      .put({ type: SET_ACTIVE_BOARDS, payload: [] })
      .dispatch({ type: CALCULATE_ACTIVE_BOARDS })
      .silentRun();
  });

  // if more put effects happen: this catches it + this checks for the order
  it('should match snapshot', () => {
    return expectSaga(activeBoardsCalculationSaga)
      .provide([[select(getMoves), movesForBoardWithThreeMovesMock]])
      .put({
        type: SET_ACTIVE_BOARDS,
        payload: activeBoardsForBoardWithThreeMovesMock,
      })
      .dispatch({ type: CALCULATE_ACTIVE_BOARDS })
      .silentRun()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});
