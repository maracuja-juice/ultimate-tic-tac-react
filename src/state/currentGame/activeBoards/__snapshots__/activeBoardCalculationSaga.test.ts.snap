// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activeBoardCalculationSaga should match snapshot 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "activeBoardsReducer/calculate-active-boards",
          "@@redux-saga-test-plan/json/function/calculateActiveBoards",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "type": "activeBoardsReducer/calculate-active-boards",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/calculateActiveBoards",
      },
      "type": "FORK",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/getMoves",
      },
      "type": "SELECT",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "activeBoardsReducer/calculate-active-boards",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "activeBoardsReducer/calculate-active-boards",
      },
      "type": "TAKE",
    },
  ],
}
`;
